File: modules/server/llm.js



(function(global) {

    var LiteGraph = global.LiteGraph;


    //node constructor class
    function MyGraphNode()
    {
    this.addInput("in","string");
    
    this.addOutput("out","string");

    let tmpl = 'type: python\ninit: "test/python_hello_world.py"'

    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    //this.container.addElement(new CustomTextarea("Parameters","parameters"))

    this.properties = {  };
    this.container.addWidget("textarea","Parameters",{ property: "parameters"})


    }

    //name to show
    MyGraphNode.title = "Generic node";

    //function to call when the node is executed
    MyGraphNode.prototype.onExecute = function()
    {
    var A = this.getInputData(0);
    if( A === undefined )
        A = 0;
    var B = this.getInputData(1);
    if( B === undefined )
        B = 0;

    }
    
    MyGraphNode.prototype.onStart = function()
    {
        console.log("started")
   
    }

    MyGraphNode.prototype.onConnectionsChange = function(dir,slot,connected,d,e)
    {
        console.log("dir: " +dir + "  slot: " + slot + "  connected: "  + connected )
        if(dir == LiteGraph.INPUT)
        {
              if(connected)
              {
                  let numInputs = this.inputs.length
                  if(slot +1 == numInputs && !!d)
                  {
                    var saved_size = this.size
                    this.addInput("N","string");
                    this.size[0] = saved_size[0]
                    this.size[1] = Math.max(this.size[1],saved_size[1])
                    //this.setSize( this.computeSize() );
                  }
              
              }
              else
              {
                    let numInputs = this.inputs.length
                    let maxInput = 0;
                    for (let i = 1; i < numInputs; i++) {
                        if(!!this.inputs[i].link)
                        {
                            maxInput = i
                        }
                    }
                     for (let i = numInputs-1; i > maxInput+1; i--) {
                             var saved_size = this.size
                            this.removeInput(i)
                            this.size[0] = saved_size[0]
                            this.size[1] = Math.max(this.size[1],saved_size[1])
                            //this.setSize( this.computeSize() );
                     
                    }
                }
        }
        if(dir == LiteGraph.OUTPUT)
        {
              if(connected)
              {
                  let numElements = this.outputs.length
                  if(slot +1 == numElements && !!d)
                  {
                    var saved_size = this.size
                    this.addOutput("N","string");
                    this.size[0] = saved_size[0]
                    this.size[1] = Math.max(this.size[1],saved_size[1])

                    //this.setSize( this.computeSize() );
                  }
                  //this.graph.change() //TODO reposition the connection in input/output node when minimized

              }
              else
              {
                setTimeout(function(){ // in case the disconnect is immediately followed by a connect
                    let numElements = this.outputs.length
                    let maxInput = 0;
                    for (let i = 1; i < numElements; i++) {
                        if((!!this.outputs[i].links) && this.outputs[i].links.length > 0)
                        {
                            maxInput = i
                        }
                    }
                     for (let i = numElements-1; i > maxInput+1; i--) {
                            var saved_size = this.size
                            this.removeOutput(i)
                            this.size[0] = saved_size[0]
                            this.size[1] = Math.max(this.size[1],saved_size[1])
                            //this.setSize( this.computeSize() );

                    }
               }.bind(this))
               }
        }
    }
    
    //register in the system
    LiteGraph.registerNodeType("graph/generic_node", MyGraphNode );

    /* ********************** ******************* */
    //node constructor class
    function MyLLMCallNode()
    {
    this.addInput("in","string");

    this.addOutput("out","string");
    this.properties = {conf:"",subtype:"stateless",template:""  };
    let tmpl = 'type: python\ninit: "test/python_hello_world.py"'

    this.addWidget("text","Config","", { property: "conf"});
    this.addProperty("subtype", "stateless", "enum", { values: ["stateless","stateful"]  });
    this.addWidget("combo","subtype","stateless",null, { property: "subtype", values: ["stateless","stateful"] } );
    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    //this.container.addElement(new CustomTextarea("Parameters","parameters"))

    
    this.container.addWidget("textarea","template",{ property: "template"})
    }

    //name to show
    MyLLMCallNode.title = "LLM Call";



    MyLLMCallNode.prototype.onConnectionsChange = MyGraphNode.prototype.onConnectionsChange

    //register in the system
    LiteGraph.registerNodeType("llm/llm_call", MyLLMCallNode );



    //node constructor class
    function MyWatchNode()
    {
    this.addInput("in","string");
    
    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    this.container.addWidget("textarea","",{ property: "parameters"})
    
    var tmpl = ""
    this.properties = { };
    }

    //name to show
    MyWatchNode.title = "Watch";

    //function to call when the node is executed
    MyWatchNode.prototype.onExecute = function()
    {
    var A = this.getInputData(0);
    this.container.setValue("parameters",A)
    }

    //register in the system
    LiteGraph.registerNodeType("llm/watch", MyWatchNode );




    //node constructor class
    function MyInputNode()
    {
    this.addOutput("out","string");

    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    this.container.addWidget("textarea","Parameters",{ property: "parameters"})


    this.properties = {  };
    }

    //name to show
    MyInputNode.title = "Text Input";

    //register in the system
    LiteGraph.registerNodeType("llm/input", MyInputNode );

    
    /****
     * 
     * 
     * 
     * LIST NODE
     * 
     * */
    function MyListNode()
    {
    this.addOutput("out","string");
    this.addInput("in","string");

    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    this.container.addWidget("list","Parameters",{ property: "parameters"})
    //this.container.addWidget("textarea","Parameters",{ property: "parameters"})
    this.properties = {  };
    }

    //name to show
    MyListNode.title = "List";

    //register in the system
    LiteGraph.registerNodeType("llm/list", MyListNode );

        /****
     * 
     * 
     * 
     * LIST NODE
     * 
     * */
    function MyChatHistoryNode()
    {
    this.addOutput("out","string");
    this.addInput("in","string");

    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    var listWidget = this.container.addWidget("list","Parameters",{ property: "parameters"})
    listWidget.makeCellName = function(index)
    {
        if(index == 0)
        {
                return "system";
        }
        else if(index % 2)
        {
                return "user";
        }
        else
        {
                return "assistant"
        }
        return index+1
    }
    listWidget.reset()
    //this.container.addWidget("textarea","Parameters",{ property: "parameters"})
    this.properties = {  };
    }

    //name to show
    MyChatHistoryNode.title = "Chat History";

    //register in the system
    LiteGraph.registerNodeType("llm/chat_history", MyChatHistoryNode );
    
    MyChatHistoryNode.prototype.onStart = function()
    {
        this.saved_history = JSON.parse(JSON.stringify(this.properties["parameters"]))
        if((this.saved_history.length % 2) != 1)
        {
            this.saved_history.push("")   
        }
    }
    
    MyChatHistoryNode.prototype.onExecute = function()
    {
        var A = this.getInputData(0);
        if( A !== undefined )
        {
            var new_history = JSON.parse(JSON.stringify(this.saved_history))
            var message_index = new_history.length-1
            new_history[message_index] =this.saved_history[message_index] + A
            this.container.setValue("parameters",new_history)
        }
    }
    

    /* ----------------- -------------- */
    //node constructor class
    function MyVirtualSourceNode()
    {
    this.addOutput("out","string");
    this.addWidget("text","identifier","", { property: "identifier"});

    this.properties = { identifier: "" };
    }

    //name to show
    MyVirtualSourceNode.title = "Virtual Source";

    //function to call when the node is executed
    MyVirtualSourceNode.prototype.onExecute = function(){}

    //register in the system
    LiteGraph.registerNodeType("graph/virtual_source", MyVirtualSourceNode );

       /* ----------------- -------------- */
    //node constructor class
    function MyVirtualSinkNode()
    {
    this.addInput("in","string");
    this.addWidget("text","identifier","", { property: "identifier"});

    this.properties = { identifier: "" };
    }

    //name to show
    MyVirtualSinkNode.title = "Virtual sink";

    //function to call when the node is executed
    MyVirtualSinkNode.prototype.onExecute = function(){}

    //register in the system
    LiteGraph.registerNodeType("graph/virtual_sink", MyVirtualSinkNode );



        //node constructor class
    function MyVariableNode()
    {

    this.addWidget("text","identifier","", { property: "identifier"});
    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    this.container.addWidget("textarea","Parameters",{ property: "parameters"})
    this.properties = { identifier: "" };
    }

    //name to show
    MyVariableNode.title = "Variable";

    //register in the system
    LiteGraph.registerNodeType("graph/variable", MyVariableNode );
    //LiteGraph.registerNodeType("llm/variable", MyVariableNode );

        //node constructor class
    function MyFileNode()
    {
    this.addOutput("out","string");
    this.addInput("in","string");

    this.addWidget("text","config","", { property: "config"});
    this.addWidget("text","filename","", { property: "filename"});

    this.properties = { filename: "", config:"" };
    }

    //name to show
    MyFileNode.title = "File";

    //register in the system
    LiteGraph.registerNodeType("tools/file", MyFileNode );
    
    
    /*
     * 
     * 
     *   AGENT NODE
     * 
     * 
     */
    //node constructor class
    function MyAgentNode()
    {
    this.addInput("query","string");
    this.addInput("LLM","string");
    this.addInput("tools","string");
    this.addInput("reflexion","string");
    
    this.addOutput("out","string");

    let tmpl = 'type: python\ninit: "test/python_hello_world.py"'

    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    //this.container.addElement(new CustomTextarea("Parameters","parameters"))

    this.properties = {  };
    this.container.addWidget("textarea","Parameters",{ property: "parameters"})
    }

    //name to show
    MyAgentNode.title = "Agent";
 
    //register in the system
    LiteGraph.registerNodeType("llm/generic_agent", MyAgentNode );


    /*

        INPUT/OUTPUT

    ********************** ******************* */

    //node constructor class
    function MyConnectionNode()
    {
        this.addInput("in","string");
        this.addOutput("out","string");

        this.properties = {conf:"",subtype:"input",template:""  };


        this.addProperty("subtype", "input", "enum", { values: ["input","output"]  });
        this.addWidget("combo","subtype","input",null, { property: "subtype", values: ["input","output"] } );
    }
    //name to show
    MyConnectionNode.title = "Hierarchical connection";
    MyConnectionNode.prototype.onConnectionsChange = MyGraphNode.prototype.onConnectionsChange
    //register in the system
    LiteGraph.registerNodeType("graph/connection", MyConnectionNode );


    /*

        WEB SCRAPER

    */
        //node constructor class
    function MyScraperNode()
    {
    this.addOutput("out","string");

    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    this.container.addWidget("text_input","URL",{ property: "address"})
    
    this.properties = { address: "" };
    }

    //name to show
    MyScraperNode.title = "Web scraper";

    //register in the system
    LiteGraph.registerNodeType("tools/web_scraper", MyScraperNode );

    /*

        PDF parser

    */
        //node constructor class
    function MyPDFNode()
    {
    this.addOutput("out","string");

    this.container = new DivContainer(this)
    this.addCustomWidget( this.container);
    this.container.addWidget("text_input","file",{ property: "address"})

    this.properties = { address: "" };
    }

    //name to show
    MyPDFNode.title = "PDF Parser";

    //register in the system
    LiteGraph.registerNodeType("tools/pdf_parser", MyPDFNode );


})(this);

File: modules/gui_nodes/llm.py
import yaml

class GuiNodeParser:
    def parse_generic(self ,old_config ,links):
        new_config = {}
        new_config["type"] = old_config["type"]
        properties = old_config.get("properties", {})
        parameters = properties.get("parameters", {})
        parameters = yaml.safe_load(parameters)

        for i, el in enumerate(parameters.get("init" ,[])):
            if isinstance(el, dict):
                if len(el) > 0:
                    val = list(el.keys())
                    parameters["init"][i] = "{" + str(val[0]) + "}"
                else:
                    parameters["init"][i] = "{}"
                pass

        if parameters:
            for vel in parameters:
                new_config[vel] = parameters[vel]


        old_inputs = old_config.get("inputs", [])
        new_inputs = [str(el["link"]) if el["link"] else None for el in old_inputs]
        new_inputs = [links[el] if el else None for el in new_inputs]
        new_inputs = [str(el[1]) + "[" + str(el[2]) + "]" if el else None for el in new_inputs]
        val_exec = []
        for vel in new_inputs:
            if not vel:
                break
            val_exec.append(vel)
        new_config["exec"] = val_exec
        return new_config

    def _calc_exec(self ,old_inputs ,links):

        new_inputs = [str(el["link"]) if el["link"] else None for el in old_inputs]
        new_inputs = [links[el] if el else None for el in new_inputs]
        new_inputs = [str(el[1]) + "[" + str(el[2]) + "]" if el else None for el in new_inputs]
        val_exec = []
        for vel in new_inputs:
            if not vel:
                break
            val_exec.append(vel)
        return val_exec

    def parse_generic_agent(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "agent"
        properties = old_config.get("properties", {})
        parameters = properties.get("parameters", {})
        parameters = yaml.safe_load(parameters)

        for i, el in enumerate(parameters.get("init" ,[])):
            if isinstance(el, dict):
                if len(el) > 0:
                    val = list(el.keys())
                    parameters["init"][i] = "{" + str(val[0]) + "}"
                else:
                    parameters["init"][i] = "{}"
                pass

        if parameters:
            for vel in parameters:
                new_config[vel] = parameters[vel]

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs ,links)
        return new_config

    def parse_llm_call(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "stateless"
        properties = old_config.get("properties", {})
        subtype = properties.get("subtype", "stateless")
        if(subtype == "stateful"):
            new_config["type"] = "stateful"
        template = properties.get("template", "")

        conf = properties.get("conf", "")
        conf = yaml.safe_load(conf)
        if conf:
            new_config["conf"] = conf
        new_config["init"] = [template]

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs ,links)
        return new_config

    def parse_file(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "file"
        properties = old_config.get("properties", {})
        template = properties.get("filename", "")
        new_config["init"] = [template]
        conf = properties.get("config", "")
        conf = yaml.safe_load(conf)
        if conf:
            new_config["conf"] = conf
        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs ,links)
        return new_config

    def parse_list(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "list"
        properties = old_config.get("properties", {})
        template = properties.get("parameters", [])
        new_config["init"] = template

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs ,links)
        return new_config

    def parse_connection(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "copy"
        new_config["conf"] = {"subtype" :"input"}
        properties = old_config.get("properties", {})
        subtype = properties.get("subtype", "input")
        if subtype == "output":
            new_config["conf"]["subtype"] = "output"

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs ,links)
        return new_config

    def parse_watch(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "copy"

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs ,links)
        return new_config

    def parse_scraper(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "exec"

        properties = old_config.get("properties", {})
        address = properties.get("address")
        new_config["init"] = ["python3", "extras/scraper/scrape.py", address]

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs ,links)
        return new_config

    def parse_pdf(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "exec"

        properties = old_config.get("properties", {})
        address = properties.get("address")
        new_config["init"] = ["python3", "extras/parse_pdf.py", address]

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs ,links)
        return new_config

    def parse_history(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "list"
        properties = old_config.get("properties", {})
        template = properties.get("parameters", [])
        t = ""
        for i ,v in enumerate(template):
            if i == 0:
                t += "{p:system}\n" + v + "{p:eom}\n\n"
            elif i % 2:
                t += "{p:user}\n" + v + "{p:eom}\n\n"
            else:
                t += "{p:assistant}\n" + v + "{p:eom}\n\n"

        new_config["init"] = [t]

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = []
        return new_config

    def parse_input(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "constant"
        properties = old_config.get("properties", {})
        parameters = properties.get("parameters", "")
        new_config["init"] = [parameters]

        old_inputs = old_config.get("inputs", [])
        new_inputs = [str(el["link"]) if el["link"] else None for el in old_inputs]
        new_inputs = [links[el] if el else None for el in new_inputs]
        new_inputs = [str(el[1]) + "[" + str(el[2]) + "]" if el else None for el in new_inputs]
        val_exec = []
        for vel in new_inputs:
            if not vel:
                break
            val_exec.append(vel)
        new_config["exec"] = val_exec
        return new_config

    def parse_python(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "python"
        properties = old_config.get("properties", {})
        parameters = properties.get("parameters", "")
        new_config["init"] = [parameters]

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs, links)
        return new_config

    def parse_tts(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "tts"
        properties = old_config.get("properties", {})
        language = properties.get("language", "it")
        new_config["conf"] = {"lang": language}

        old_inputs = old_config.get("inputs", [])
        new_config["exec"] = self._calc_exec(old_inputs, links)
        return new_config

    def parse_variable(self ,old_config ,links):
        new_config = {}
        new_config["type"] = "variable"
        properties = old_config.get("properties", {})
        parameters = properties.get("parameters", "")

        new_config["conf"] = {"name" :properties["identifier"] ,"value" :properties["parameters"]}

        old_inputs = old_config.get("inputs", [])
        new_inputs = [str(el["link"]) if el["link"] else None for el in old_inputs]
        new_inputs = [links[el] if el else None for el in new_inputs]
        new_inputs = [str(el[1]) + "[" + str(el[2]) + "]" if el else None for el in new_inputs]
        val_exec = []
        for vel in new_inputs:
            if not vel:
                break
            val_exec.append(vel)
        new_config["exec"] = val_exec
        return new_config

    def parse_virtual(self ,old_config ,links):
        new_config = {}
        new_config["type"] = old_config["type"].split("/")[-1]
        properties = old_config.get("properties", {})
        parameters = properties.get("identifier", "")
        new_config["conf"] = {"name" :parameters}

        old_inputs = old_config.get("inputs", [])
        new_inputs = [str(el["link"]) if el["link"] else None for el in old_inputs]
        new_inputs = [links[el] if el else None for el in new_inputs]
        new_inputs = [str(el[1]) + "[" + str(el[2]) + "]" if el else None for el in new_inputs]
        val_exec = []
        for vel in new_inputs:
            if not vel:
                break
            val_exec.append(vel)
        new_config["exec"] = val_exec
        return new_config

    def parse_node(self ,old_config ,links):

        new_config = {}
        new_config["type"] = old_config["type"]
        node_type = old_config["type"].split("/")[-1]
        if node_type in ["generic_node"]:
            return self.parse_generic(old_config ,links)
        if old_config["type"] in ["llm/input"]:
            return self.parse_input(old_config ,links)
        if node_type in ["virtual_sink" ,"virtual_source"]:
            return self.parse_virtual(old_config ,links)
        if old_config["type"].startswith("llm/watch"):
            return None
        if node_type in ["variable"]:
            return self.parse_variable(old_config ,links)
        if old_config["type"].startswith("llm/llm_call"):
            return self.parse_llm_call(old_config ,links)
        if node_type in ["file"]:
            return self.parse_file(old_config ,links)
        if old_config["type"].startswith("llm/list"):
            return self.parse_list(old_config ,links)
        if old_config["type"].startswith("llm/chat_history"):
            return self.parse_history(old_config ,links)
        if old_config["type"].startswith("llm/generic_agent"):
            return self.parse_generic_agent(old_config ,links)
        if node_type in ["connection"]:
            return self.parse_connection(old_config ,links)
        if node_type in ["web_scraper"]:
            return self.parse_scraper(old_config ,links)
        if node_type in ["pdf_parser"]:
            return self.parse_pdf(old_config ,links)
        if node_type in ["watch"]:
            return self.parse_watch(old_config ,links)
        if node_type in ["python_sandbox"]:
            return self.parse_python(old_config ,links)
        if node_type in ["tts"]:
            return self.parse_tts(old_config ,links)
        return None

    def postprocess_nodes(self,new_nodes):
        virtual_sinks = [new_nodes[el] for el in new_nodes if new_nodes[el]["type"] == "virtual_sink"]
        virtual_sources = [new_nodes[el] for el in new_nodes if new_nodes[el]["type"] == "virtual_source"]
        for el in virtual_sources:
            identifier = el["conf"]["name"]
            compatible_sinks = [e for e in new_nodes if new_nodes[e]["type"] == "virtual_sink" and new_nodes[e]["conf"]["name"] == identifier]
            el["type"] = "copy"
            el["exec"] = [compatible_sinks[0]]

        for el in virtual_sinks:
            el["type"] = "copy"

        agents = [(el, new_nodes[el]) for el in new_nodes if new_nodes[el]["type"] == "agent"]
        for agent_name, agent in agents:
            deps = agent["exec"]
            LLM = deps[1].split("[")[0]
            node = new_nodes[LLM]
            node["exec"] = [agent_name + "[1]"]
            tools = deps[2].split("[")[0]
            node = new_nodes[tools]
            node["exec"] = [agent_name + "[2]"]
            if len(deps) > 3:
                reflex = deps[3].split("[")[0]
                node = new_nodes[reflex]
                node["exec"] = [agent_name + "[3]"]